app/routes/reorder/_data-sources/track.server.js

```diff
@@ -0,0 +1,78 @@
+import { prisma } from "~/db.server";
+
+export const getTracks = () =>
+  prisma.track.findMany({ orderBy: { trackNumber: "asc" } });
+
+export const moveTrackUp = async (id) => {
+  const fromTrackNumber = await getTrackNumber(id);
+
+  if (fromTrackNumber === 1) return;
+
+  const toTrackNumber = fromTrackNumber - 1;
+  const toTrackId = await getTrackIdFromTrackNumber(toTrackNumber);
+
+  await swapTrackNumbers({
+    fromTrackId: id,
+    fromTrackNumber,
+    toTrackId,
+    toTrackNumber,
+  });
+};
+
+export const moveTrackDown = async (id) => {
+  const fromTrackNumber = await getTrackNumber(id);
+
+  const {
+    _max: { trackNumber: maxTrackNumber },
+  } = await prisma.track.aggregate({
+    _max: { trackNumber: true },
+  });
+
+  if (fromTrackNumber === maxTrackNumber) return;
+
+  const toTrackNumber = fromTrackNumber + 1;
+  const toTrackId = await getTrackIdFromTrackNumber(toTrackNumber);
+
+  await swapTrackNumbers({
+    fromTrackId: id,
+    fromTrackNumber,
+    toTrackId,
+    toTrackNumber,
+  });
+};
+
+const getTrackIdFromTrackNumber = async (trackNumber) => {
+  const { id } = await prisma.track.findFirst({
+    where: { trackNumber },
+    select: { id: true },
+  });
+  return id;
+};
+
+const getTrackNumber = async (id) => {
+  const { trackNumber } = await prisma.track.findUnique({
+    where: { id },
+    select: { trackNumber: true },
+  });
+  return trackNumber;
+};
+
+const swapTrackNumbers = async ({
+  fromTrackId,
+  fromTrackNumber,
+  toTrackId,
+  toTrackNumber,
+}) => {
+  await prisma.$transaction(
```

Comment:
Let's use a transaction so the whole thing can be rolled back if it fails, and commits all at once.

Reply:
@gsong why do you need to swap? I mean, you have the `from` and `to` trackNumbers, couldn't you just set them up?

Reply:
Well, since `trackNumber` has to be unique, and I can swap the track numbers with one SQL call (at least not that I can think of), we need to temporarily set one of them to something else, otherwise we would violate the unique constraint.

Reply:
Ah that's true, I missed that, thanks